c/src/lib.rs

```diff
@@ -2,16 +2,90 @@
+#[derive(Clone, Copy)]
+#[repr(transparent)]
+pub struct FoxgloveServerCapability {
+    pub flags: u8,
+}
+/// Allow clients to advertise channels to send data messages to the server.
+pub const FOXGLOVE_SERVER_CAPABILITY_CLIENT_PUBLISH: u8 = 1 << 0;
+/// Allow clients to subscribe and make connection graph updates
+pub const FOXGLOVE_SERVER_CAPABILITY_CONNECTION_GRAPH: u8 = 1 << 1;
+/// Allow clients to get & set parameters.
+pub const FOXGLOVE_SERVER_CAPABILITY_PARAMETERS: u8 = 1 << 2;
+/// Inform clients about the latest server time.
+///
+/// This allows accelerated, slowed, or stepped control over the progress of time. If the
+/// server publishes time data, then timestamps of published messages must originate from the
+/// same time source.
+pub const FOXGLOVE_SERVER_CAPABILITY_TIME: u8 = 1 << 3;
+/// Allow clients to call services.
+pub const FOXGLOVE_SERVER_CAPABILITY_SERVICES: u8 = 1 << 4;
+
+bitflags! {
+    #[derive(Clone, Copy, PartialEq, Eq)]
+    struct FoxgloveServerCapabilityBitFlags: u8 {
+        const ClientPublish = FOXGLOVE_SERVER_CAPABILITY_CLIENT_PUBLISH;
+        const ConnectionGraph = FOXGLOVE_SERVER_CAPABILITY_CONNECTION_GRAPH;
+        const Parameters = FOXGLOVE_SERVER_CAPABILITY_PARAMETERS;
+        const Time = FOXGLOVE_SERVER_CAPABILITY_TIME;
+        const Services = FOXGLOVE_SERVER_CAPABILITY_SERVICES;
+    }
+}
+
+impl FoxgloveServerCapabilityBitFlags {
+    fn iter_websocket_capabilities(self) -> impl Iterator<Item = foxglove::websocket::Capability> {
+        self.iter_names().filter_map(|(_s, cap)| match cap {
+            FoxgloveServerCapabilityBitFlags::ClientPublish => {
+                Some(foxglove::websocket::Capability::ClientPublish)
+            }
+            FoxgloveServerCapabilityBitFlags::ConnectionGraph => {
+                Some(foxglove::websocket::Capability::ConnectionGraph)
+            }
+            FoxgloveServerCapabilityBitFlags::Parameters => {
+                Some(foxglove::websocket::Capability::Parameters)
+            }
+            FoxgloveServerCapabilityBitFlags::Time => Some(foxglove::websocket::Capability::Time),
+            FoxgloveServerCapabilityBitFlags::Services => {
+                Some(foxglove::websocket::Capability::Services)
+            }
+            _ => None,
+        })
+    }
+}
+
+impl From<FoxgloveServerCapability> for FoxgloveServerCapabilityBitFlags {
+    fn from(bits: FoxgloveServerCapability) -> Self {
+        Self::from_bits_retain(bits.flags)
+    }
+}
+
 #[repr(C)]
 pub struct FoxgloveServerOptions<'a> {
     pub name: *const c_char,
     pub host: *const c_char,
     pub port: u16,
     pub callbacks: Option<&'a FoxgloveServerCallbacks>,
+    pub capabilities: FoxgloveServerCapability,
+    pub supported_encodings: *const *const c_char,
+    pub supported_encodings_count: usize,
+}
+
+#[repr(C)]
+pub struct FoxgloveClientChannel<'a> {
+    pub id: u32,
+    pub topic: &'a c_char,
```

Comment:
So weird to see these as references instead of pointers

Reply:
Yeah, I guess I was hoping that it would be helpful to have the nonnull annotation on the C header side. In practice I'm not sure it has really helped that much...can remove.